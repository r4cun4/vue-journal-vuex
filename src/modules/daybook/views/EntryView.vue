import { defineAsyncComponent } from 'vue';
<template>
    <template v-if="entry">
        <div class="entry-title d-flex justify-content-between p-2">
            <div>
                <span class="text-success fs-3 fw-bold">{{ day }}</span>
                <span class="mx-1 fs-3">{{ month }}</span>
                <span class="mx-1 fs-3 fw-light">{{ yearDay }}</span>
            </div>

            <div>
                <button class="btn btn-danger mx-2">
                    Borrar
                    <i class="fa fa-trash alt"></i>
                </button>

                <button class="btn btn-primary">
                    Borrar
                    <i class="fa fa-upload"></i>
                </button>
            </div>
        </div>

        <hr>
        <div class="d-flex flex-column px-3 h-75">
            <textarea
                v-model="entry.text"
                placeholder="¿Qué sucedió hoy?"
            ></textarea>
        </div>

        <img
            src="https://www.imagenesbonitasname.com/covers/thumbs/gatito-imagen-tierna.jpg"
            alt="entry-picture"
            class="img-thumbnail">

    </template>

    <TheFab
        icon="fa-save"
    />
</template>

<script>
import { defineAsyncComponent } from 'vue'
import { mapGetters } from 'vuex'; // computed

import getDayMonthYear from '../components/helpers/getDayMonthYear'

export default {
    props: {
        id: {
            type: String,
            required: true
        }
    },
    components: {
        TheFab: defineAsyncComponent(() => import('../components/TheFab.vue'))
    },

    data() {
        return {
            entry: null
        }
    },

    computed: {
        ...mapGetters('journalModule', ['getEntriesById']),
        day() {
            const { day } = getDayMonthYear( this.entry.date )
            return day
        },
        month() {
            const { month } = getDayMonthYear( this.entry.date )
            return month
        },
        yearDay() {
            const { yearDay } = getDayMonthYear( this.entry.date )
            return yearDay
        }
    },

    methods: {
        loadEntry() {
            const entry = this.getEntriesById( this.id )
            if ( !entry ) return this.$router.push({ name: 'no-entry' })

            this.entry = entry
        }
    },

    created() {
        this.loadEntry()
    },

    watch: {
        id() {
            this.loadEntry()
        }
    }
}
</script>

<style lang="scss" scoped>
textarea {
    font-size: 20px;
    border: none;
    height: 100%;

    &:focus {
        outline: none;
    }
}

img {
    width: 200px;
    position: fixed;
    bottom: 150px;
    right: 20px;
    box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
}
</style>