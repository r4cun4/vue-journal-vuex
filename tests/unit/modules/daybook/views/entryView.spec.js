import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'

import Swal from 'sweetalert2'

import journalModule from '@/modules/daybook/store/journal'
import { journalState } from '../../../mock-data/test-journal-state'

import EntryView from '@/modules/daybook/views/EntryView'

const createVuexStore = ( initialState ) =>
    createStore({
        modules: {
            journalModule: {
                ...journalModule,
                state: { ...initialState }
            }
        }
    })

jest.mock('sweetAlert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))

describe('Pruebas en el EntryView', () => {

    const store = createVuexStore( journalState )
    store.dispatch = jest.fn()
    const mockRouter = {
        push: jest.fn()
    }

    let wrapper

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount( EntryView, {
            props: {
                id: '-NPSw_XMR-dMQr7gfm0g'
            },
            global: {
                mocks: {
                    $router: mockRouter 
                },
                plugins: [ store ]
            }
        })
    })

    test('debe de sacar al usuario porque el id no existe', () => {

        const wrapper = shallowMount( EntryView, {
            props: {
                id: 'Este ID no existe en el STORE'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })

        expect( mockRouter.push ).toHaveBeenCalledWith({ name: 'no-entry' })

    })

    test('debe de mostrar la entrada correctamente', () => {

        expect(wrapper.html()).toMatchSnapshot()
        expect(mockRouter.push).not.toHaveBeenCalled()

    })

    test('debe de borrar la entrada y salida', ( done ) => {

        Swal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed: true}))

        wrapper.find('.btn-danger').trigger('click')

        expect( Swal.fire ).toHaveBeenCalledWith({
            title: '¿Está seguro?',
            text: 'Una vez borrado, no se puede recuperar',
            showDenyButton: true,
            confirmButtonText: 'Si, estoy seguro'
        })

        setTimeout( () => {

            expect( store.dispatch ).toHaveBeenCalledWith('journalModule/deleteEntry', '-NPSw_XMR-dMQr7gfm0g')
            expect( mockRouter.push ).toHaveBeenCalled()
            done()

        }, 1)
    })

})