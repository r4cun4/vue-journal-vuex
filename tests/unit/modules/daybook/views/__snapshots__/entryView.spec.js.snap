// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">1</span><span class="mx-1 fs-3">Marzo</span><span class="mx-1 fs-3 fw-light">2023, Miércoles</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué sucedió hoy?"></textarea></div>
<!--v-if-->
<!--v-if-->
<the-fab-stub icon="fa-save"></the-fab-stub>
`;
